### ISSUE TITLE
Access to current controller in `on_server_error` handler

### ISSUE THREAD

**Issue by benarmston** [JSONAPI-Resources/jsonapi-resources#734]
I'm attempting to notify airbrake of exceptions raised when jsonapi-resources processes a request.  There are two separate paths that exceptions raised by jsonapi-resources can take.
1. Exceptions raised by `render_errors` or `render_results` which are handled by `handle_exceptions`.
2. Exceptions raised by `OperationsProcessor` which are handled by the `server_error_callbacks` (a.k.a.  `on_server_error` callbacks).

In the first of those cases, the exception handler has access to the controller instance created for handling the request.  This means that it is possible to include data about the request that caused the exception, such as the session, any parameters, etc..

In the second of those cases, the exception handler does not have access to the controller instance.  So whilst I can report the exception, a lot of useful information is not available.

I can work around this by adding `Exception` to the whitelist of exceptions, then all exceptions are handled by `handle_exceptions`.  But I suspect that I'm missing out on useful error handling by jsonapi-resources.  If I'm not missing out on anything, then I'm not sure why we have `server_error_callbacks` and `on_server_error`.

Assuming that `server_error_callbacks` is useful, would it not make sense to either:
1. Pass a copy of the current controller instance to the `Request` instance, so that it can be included in the call to the `on_server_error` callbacks; or
2. Have `process_request` augment the `server_error_callbacks` so that they are passed the current controller instance.

Effectively, these are the same solution, but the second avoids providing the controller instance to the Request instance, which may be undesirable.

Or am I missing something very obvious here.

**Comment by mmartinson** [comment]
This is definitely a valid issue and could use some attention, and I think you description captures the problem exactly. There could be some changes made here but caution is warranted because bugs in error handling code can pass unnoticed for a long time. 

A couple of questions and considerations:

1) If the callbacks are defined as instance methods in the controller and we can call `self` to get data from the request, how should this be represented in block callbacks, should controller instance be available in blocks at all, and if not, is it useful to support them? The current implementation passes both styles of callback as procs into the request. It would be possible to pass a collection of symbols instead for instance methods that should be called by the request. This would still involve passing the controller instance to the request, but it could at least be maintained as private state and not exposed

2) I like the simplicity of option 2, but it might be nice to make the changes so callbacks aren't defined at the class level. I think this might be more intuitive than the weirdness of receiving the controller instance as an argument to a class method. Would love some feedback on this point.

3) If the callbacks are going to be switched to instance-level, lets add a deprecation notice for class methods and leave them until the next release.

**Comment by xdmx** [comment]
I'm facing this same need. I've tried to add the `on_server_error` sample in the README but it complains that there is no `notify_airbrake` method, maybe it's because I'm including `JSONAPI::ActsAsResourceController` under `ActionController::API`

So far I've solved by sending just the error, but having all the request info would be very useful

```
module V1
  class BaseController < ActionController::API
    include JSONAPI::ActsAsResourceController

    on_server_error do |error|
      Airbrake.notify(error)
    end
  end
end
```

**Comment by seanmcc** [comment]
I realize this is a bit late, but I did figure out a hacky work around to get access to the request.
```
error.bindings.detect { |b| b.local_variables.include?(:request) }&.local_variable_get(:request)
```
That steps through the errors bindings and finds the context where request is present.

**Comment by tindron** [comment]
@seanmcc Where are you getting the bindings method on your error from?  I don't have one of those.

**Comment by tommy-russoniello** [comment]
Fixed in 4c856ef59d80afd2484eedb20e350330688d04e5.
